api/client.go

```diff
@@ -418,13 +448,90 @@ func NewClient(c *Config) (*Client, error) {
 		config: c,
 	}
 
-	if token := os.Getenv(EnvVaultToken); token != "" {
+	// determine how to get a token
+	// order of precedence could change
+	tokenSinkPath := ""
+	switch {
+	case os.Getenv(EnvVaultToken) != "":
+		client.token = os.Getenv(EnvVaultToken)
+	case c.TokenFileSinkPath != "":
+		tokenSinkPath = c.TokenFileSinkPath
+	case c.AgentAddress != "":
+		tokenSinkPath, err = client.getSinkPathFromAgent(c.AgentAddress)
+		if err != nil {
+			return nil, errwrap.Wrapf(fmt.Sprintf("failed to determine tokenSinkPath given the provided agent address %q {{err}}", c.AgentAddress), err)
+		}
+	}
+
+	// set and poll token from file sink if it is available
+	if tokenSinkPath != "" {
+		token, err := readAgentTokenFromFile(tokenSinkPath)
+		if err != nil {
+			return nil, err
+		}
 		client.token = token
+		// poll file for updates
+		client.pollFileForToken(tokenSinkPath)
 	}
 
 	return client, nil
 }
 
+// contacts agent for a file sink path
+func (c *Client) getSinkPathFromAgent(agentAddress string) (string, error) {
+	uri := fmt.Sprintf("%s%s", agentAddress, consts.AgentPathFileSinks)
+	if c.config.AgentSinkName != "" {
+		uri = fmt.Sprintf("%s?sinkName=%s", uri, url.QueryEscape(c.config.AgentSinkName))
+	}
+
+	response, err := c.config.HttpClient.Get(uri)
```

Comment:
I haven't done a lot of web stuff in Go, so take this with a grain of salt, but here's how I might write the above:

```
	uri, err := url.Parse(agentAddress)
	if err != nil {
		return "", err
	}
	uri.Path = consts.AgentPathFileSinks
	if c.config.AgentSinkName != "" {
		uri.RawQuery = url.Values{"sinkName": c.config.AgentSinkName}.Encode()
	}
	response, err := c.config.HttpClient.Get(uri.String())
```

I like to parse the agentAddress so that if there's a problem with it we discover it early, allowing us to give clearer error messages to the user.  An additional optional step I didn't add would be to verify that the url's Path component is empty.

I also prefer to avoid manual string manipulation of the URL when possible, though that's probably just my own idiosyncrasy.